### DISCUSSION TITLE
mesh network setup in Zephyr nrf52840

### DISCUSSION THREAD

**Discussion by JasmineRandhawa** [zephyrproject-rtos/zephyr#28985] [Q&A] [answered]
I have one query. I want to set up a small mesh network using three nrf52840 boards say B1,B2 and B3 . I wish to track proximity of B2 and B3 on my main board B1.  So should I be programming B2 and B3 as beacons ? And how do I setup mesh network do I need BLE pairing?

**Comment by carlescufi** [comment]
@trond-snekvik FYI

**Comment by trond-snekvik** [comment, marked as answer]
*Response from slack:*

Having B2 and B3 send beacons should work. You'll get the RSSI through the scan callback on B1. RSSI is pretty inaccurate though. It's heavily affected by obstacles, and will fluctuate a lot from packet to packet. In an open environment, you'll be able to tell distance with an accuracy of a couple of meters if you calibrate properly.
Beacon sending and receiving is enough to track proximity, but if you want more advanced communication between the boards, you should set up a mesh network. The right way to set up the network is through provisioning, but it's possible to assign addresses and keys directly to the devices through the local apis. This is considered an insecure solution though, so it's only suitable for small projects that won't do anything security critical.

**Reply by JasmineRandhawa** [reply]
@trond-snekvik I deeply appreciate and Thanks  for taking out time for helping me out with this...

I just want to confirm to see I am getting this right...

so one way to setup mesh network is that  B1 would be first establishing a connection via Bluetooth with beacons B2 and B3 and then after connection is established they can disconnect and form a mesh network right? And then they can keep sending rssi values to B1 ...B1 can read the values and calculate their proximity...
So bluetooth is just to set up the initial connection and later does not do anything once mesh is set up. 

And another way is B1 can directly create a mesh network through provisioning By hard coding the ids of B2 and B3 and wont be needing Bluetooth connection to be established

**Reply by trond-snekvik** [reply]
Hold on, I think we're not quite in sync here. 
There are essentially two ways to solve the usage scenario you described first:
1. Don't use Bluetooth Mesh at all, just send advertising beacons from B2 and B3, and perform scanning on B1 to read out the contents and RSSI of the beacons to track their proximity. The [iBeacon sample](https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/bluetooth/ibeacon) is an excellent starting point for this.
2. Use Bluetooth Mesh to communicate. Bluetooth Mesh requires a lot more effort, but offers two way communication, encryption and authentication. To establish a Bluetooth Mesh network, you need to provision the devices into the same network. The recommended way to do this is to provision the devices, for instance through the [nRF Mesh applications for Android and iOS](https://docs.zephyrproject.org/latest/guides/bluetooth/bluetooth-tools.html#mobile-applications). You can also perform provisioning with hardcoded values on each device, but this is insecure, and should never be done in projects that will leave the building. Once you have created a Mesh network, you should send model messages with TTL=0 from B2 and B3 to a predetermined mesh group address B1 subscribes to, or directly to B1's unicast address. There are no standard mesh models that are suitable for this, so you'll have to make your own vendor models, like the [onoff-vnd sample](https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/boards/nrf/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c). 

Unless you really need two way communication and authentication, I'd strongly recommend just using iBeacon or a similar approach. Bluetooth Mesh is a very powerful protocol, but has a significant cost of entry for new users.

**Reply by JasmineRandhawa** [reply]
Many thanks @trond-snekvik for providing clarification. I would program b2 and b3 as ibeacon. For b1 is there any sample in zephyr that create mesh network through provisioning by hard coding values of the beacons. I would probably be needing only one way communication from beacons to my main board B1 and B1 is just reading rssi value and not sending any to beacons.
I am doing this as purpose of learning as a student as i found embedded programming interesting. 
Thanks a lot again

**Reply by trond-snekvik** [reply]
The mesh provisioner sample does this.

**Reply by JasmineRandhawa** [reply]
Thank you @trond-snekvik
